atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CellRangeDeleter.java

```diff
@@ -108,29 +107,45 @@ public String toString() {
         }
     }
 
-    private void insertRangeTombstones(CassandraClient client, Map<Cell, Long> maxTimestampExclusiveByCell,
-            TableReference tableRef, boolean deleteSentinel, long rangeTombstoneCassandraTs) throws TException {
-        MutationMap mutationMap = new MutationMap();
+    /**
+     * If we are deleting inclusive, we must first delete exclusive. This is because although we delete at consistency
+     * all, this doesn't mean that the write fails if there are nodes down; it means that the write partially applies
+     * if there are nodes down. This leads to a scenario where data can come back from the dead in some situations.
+     * This can only affect us if there is no covering tombstone, so we make sure that we've deleted all historical
+     * versions before deleting the latest cell (which is an Atlas level tombstone).
+     */
+    private void insertRangeTombstones(
+            CassandraClient client, Map<Cell, TimestampRangeDelete> deletes,
+            TableReference tableRef, long rangeTombstoneCassandraTs) throws TException {
+        insertTombstones(client, deletes, tableRef, rangeTombstoneCassandraTs, TimestampRangeDelete::timestamp);
+        insertTombstones(client, Maps.filterValues(deletes, TimestampRangeDelete::endInclusive),
+                tableRef, rangeTombstoneCassandraTs, delete -> delete.maxTimestampToDelete() + 1);
```

Comment:
Is this an exclusive bound (maxTimestampToDelete)? If so, can we make this clearer?

Reply:
I've renamed the parameter in `insertTombstones` to be very clear that the function returns the _exclusive_ max timestamp to delete.